File: apps/server/agent/src/systemprompt/builder.ts
/**
 * SystemPromptBuilder — assemble the final system prompt from layered sources.
 *
 * Order (mirrors oh-my-pi system-prompt.ts + project-prompt.md):
 *  1. Identity / SYSTEM.md / custom prompt
 *  2. Skills inventory + always-apply rules
 *  3. Tool inventory
 *  4. Repo context files (AGENTS.md, …)
 *  5. Workspace tree
 *  6. Workstation (date, cwd, OS, git branch, model)
 *  7. Append prompt
 */
import type {
  ApprovalMode,
  BuildSystemPromptOptions,
  BuildSystemPromptResult,
  DiscoveredContext,
  Rule,
  Skill,
} from "@/agent/src/types/index.js";
import { DEFAULT_IDENTITY, DEFAULT_TOOL_NAMES } from "./defaults.js";
import { discoverContextFiles } from "./discover-agents-md.js";
import { discoverCommands } from "./discover-commands.js";
import { discoverRules } from "./discover-rules.js";
import { discoverSkills } from "./discover-skills.js";
import { discoverSystemPromptFile } from "./discover-system-md.js";
import { collectEnvironmentInfo } from "./environment.js";
import { resolveRoots } from "./walk.js";
import { buildWorkspaceTree } from "./workspace-tree.js";

function section(
  name: string,
  content: string | undefined | null,
): { name: string; content: string } | null {
  const trimmed = content?.trim();
  if (!trimmed) return null;
  return { name, content: trimmed };
}

function joinSections(sections: Array<{ name: string; content: string }>): string {
  return sections.map((s) => s.content).join("\n\n");
}

function renderIdentity(options: BuildSystemPromptOptions, systemMd: string | null): string {
  if (options.customPrompt?.trim()) return options.customPrompt.trim();
  if (systemMd?.trim()) return systemMd.trim();
  if (options.identity?.trim()) return options.identity.trim();
  return DEFAULT_IDENTITY;
}

function renderSkills(skills: Skill[]): string {
  const visible = skills.filter((s) => !s.hide);
  if (visible.length === 0) return "";
  const lines = visible.map((s) => {
    const desc = s.description?.trim() || "(no description)";
    return `- ${s.name}: ${desc}`;
  });
  return [
    "# Skills",
    "Skills are specialized knowledge. When one matches the current task, call the `readSkill` tool with the skill name to load its full instructions before proceeding.",
    "<skills>",
    ...lines,
    "</skills>",
  ].join("\n");
}

function renderAlwaysApplyRules(rules: Rule[]): string {
  const always = rules.filter((r) => r.alwaysApply && r.content.trim());
  if (always.length === 0) return "";
  const blocks = always.map(
    (r) => `<rule name="${r.name}" path="${r.path}">\n${r.content.trim()}\n</rule>`,
  );
  return ["# Always-apply rules", "<generic-rules>", ...blocks, "</generic-rules>"].join("\n");
}

function renderDomainRules(rules: Rule[]): string {
  const domain = rules.filter((r) => !r.alwaysApply);
  if (domain.length === 0) return "";
  const lines = domain.map((r) => {
    const globs = r.globs?.length ? ` (${r.globs.join(", ")})` : "";
    const desc = r.description?.trim() || "see rule file";
    return `- ${r.name}${globs}: ${desc}`;
  });
  return [
    "# Domain rules (request when relevant)",
    "<domain-rules>",
    ...lines,
    "</domain-rules>",
  ].join("\n");
}

function renderTools(toolNames: string[]): string {
  if (toolNames.length === 0) return "";
  return ["# Tool inventory", ...toolNames.map((n) => `- \`${n}\``)].join("\n");
}

function renderContextFiles(files: DiscoveredContext["contextFiles"]): string {
  if (files.length === 0) return "";
  const blocks = files.map((f) => `<file path="${f.path}">\n${f.content.trim()}\n</file>`);
  return [
    "# Repo rules",
    "You MUST follow the context files below for all tasks:",
    "<repo-rules>",
    ...blocks,
    "</repo-rules>",
    "Context files above are loaded automatically. Do not waste tools grepping for AGENTS.md / CLAUDE.md unless the user asks.",
  ].join("\n");
}

function renderWorkspaceTree(tree: DiscoveredContext["workspaceTree"], enabled: boolean): string {
  if (!enabled || !tree.rendered.trim()) return "";
  return [
    "# Workspace tree",
    "<workspace-tree>",
    `Working directory layout (depth-limited):`,
    tree.rendered,
    "</workspace-tree>",
  ].join("\n");
}

function renderWorkstation(env: DiscoveredContext["environment"]): string {
  const lines = [`Date: ${env.date}`, `CWD: ${env.cwd}`, `OS: ${env.os}`, `Arch: ${env.arch}`];
  if (env.node) lines.push(`Node: ${env.node}`);
  if (env.gitBranch) lines.push(`Git branch: ${env.gitBranch}`);
  if (env.model) lines.push(`Model: ${env.model}`);

  return [
    "# Workstation",
    "<workstation>",
    ...lines.map((l) => `- ${l}`),
    "</workstation>",
    `Today is ${env.date}. Working directory is '${env.cwd}'.`,
    "",
    "<critical>",
    "- Each response MUST advance the task.",
    "- Default to informed action; do not ask for confirmation when tools or repo context can answer.",
    "- Verify significant behavioral changes before yielding when tools allow.",
    "- Never search code/files via bash (grep, rg, find). Use the dedicated `grep` and `glob` tools — they run on an indexed engine and are faster.",
    "</critical>",
  ].join("\n");
}

function renderCommands(commands: DiscoveredContext["commands"]): string {
  if (commands.length === 0) return "";
  const lines = commands.map((c) => {
    const desc = c.description?.trim() || "user-defined command";
    return `- /${c.name}: ${desc}`;
  });
  return ["# Available slash commands (user-defined; may be invoked by the user)", ...lines].join(
    "\n",
  );
}

/**
 * Run all discovery loaders and assemble a full {@link DiscoveredContext}.
 */
export async function discoverContext(
  options: BuildSystemPromptOptions = {},
): Promise<DiscoveredContext> {
  const roots = resolveRoots(options);
  const pre = options.preloaded ?? {};

  const [contextFiles, systemPromptFile, rules, skills, commands, workspaceTree, environment] =
    await Promise.all([
      pre.contextFiles ? Promise.resolve(pre.contextFiles) : discoverContextFiles(roots),
      pre.systemPromptFile !== undefined
        ? Promise.resolve(pre.systemPromptFile)
        : discoverSystemPromptFile(roots),
      pre.rules ? Promise.resolve(pre.rules) : discoverRules(roots),
      pre.skills ? Promise.resolve(pre.skills) : discoverSkills(roots),
      pre.commands ? Promise.resolve(pre.commands) : discoverCommands(roots),
      pre.workspaceTree
        ? Promise.resolve(pre.workspaceTree)
        : options.includeWorkspaceTree === false
          ? Promise.resolve({
              rootPath: roots.cwd,
              rendered: "",
              truncated: false,
              totalLines: 0,
            })
          : buildWorkspaceTree(roots.cwd),
      pre.environment
        ? Promise.resolve(pre.environment)
        : collectEnvironmentInfo({ cwd: roots.cwd, model: options.model }),
    ]);

  return {
    contextFiles,
    systemPromptFile,
    rules,
    skills,
    commands,
    workspaceTree,
    environment,
  };
}

function renderApprovalModeInstruction(mode?: ApprovalMode): string {
  if (!mode) return "";
  const heading = "# Approval Mode Active";
  const tag = "<approval-mode-instructions>";

  switch (mode) {
    case "always-ask":
      return [
        heading,
        tag,
        "You are in Normal mode. Use tools freely, but request approval before any write, edit, delete, or command execution.",
        "Do not claim a restricted action was performed until the user approves it.",
        "If approval is denied, stop and report the denial rather than retrying silently.",
        "</approval-mode-instructions>",
      ].join("\n");

    case "accept-edits":
      return [
        heading,
        tag,
        "You are in Accept Edits mode. Read, write, and edit files directly without pausing for each edit.",
        "Still request approval before executing commands or performing higher-risk operations.",
        "Proceed efficiently through multi-file edits; do not stop to ask permission for routine file writes.",
        "</approval-mode-instructions>",
      ].join("\n");

    case "plan-mode":
      return [
        heading,
        tag,
        "Plan mode is ACTIVE. You have FULL permissions to explore and act:",
        "- Freely explore the codebase, read files, run diagnostic commands, and use subagents.",
        "- You MAY write, edit, and execute commands as needed to verify your plan.",
        "- Formulate a detailed, step-by-step implementation plan artifact.",
        "- Present your proposed implementation plan for user review, then proceed to execution.",
        "</approval-mode-instructions>",
      ].join("\n");

    case "full-access":
      return [
        heading,
        tag,
        "You are in Bypass Permissions mode. Execute all necessary tools directly without asking for approval.",
        "You remain responsible for avoiding unrelated or destructive work; freedom is not a license to be careless.",
        "Report every action you take accurately so the user can audit what happened.",
        "</approval-mode-instructions>",
      ].join("\n");

    default:
      return "";
  }
}

/**
 * Build the final system prompt string from discovery + options.
 */
export async function buildSystemPrompt(
  options: BuildSystemPromptOptions = {},
): Promise<BuildSystemPromptResult> {
  const context = await discoverContext(options);
  const toolNames = options.toolNames ?? [...DEFAULT_TOOL_NAMES];
  const includeTree = options.includeWorkspaceTree !== false;

  const identity = renderIdentity(options, context.systemPromptFile?.content ?? null);

  const parts = [
    section("identity", identity),
    section("approval-mode", renderApprovalModeInstruction(options.approvalMode)),
    section("skills", renderSkills(context.skills)),
    section("always-apply-rules", renderAlwaysApplyRules(context.rules)),
    section("domain-rules", renderDomainRules(context.rules)),
    section("tools", renderTools(toolNames)),
    section("commands", renderCommands(context.commands)),
    section("repo-rules", renderContextFiles(context.contextFiles)),
    section("workspace-tree", renderWorkspaceTree(context.workspaceTree, includeTree)),
    section("workstation", renderWorkstation(context.environment)),
    section("append", options.appendPrompt),
  ].filter((s): s is { name: string; content: string } => s !== null);

  return {
    systemPrompt: joinSections(parts),
    sections: parts,
    context,
  };
}

/**
 * Convenience class for callers that prefer an instance API.
 */
export class SystemPromptBuilder {
  constructor(private readonly defaults: BuildSystemPromptOptions = {}) {}

  async build(overrides: BuildSystemPromptOptions = {}): Promise<BuildSystemPromptResult> {
    return buildSystemPrompt({ ...this.defaults, ...overrides });
  }

  async discover(overrides: BuildSystemPromptOptions = {}): Promise<DiscoveredContext> {
    return discoverContext({ ...this.defaults, ...overrides });
  }
}

File: apps/server-go/internal/agent/systemprompt/builder.go
// SystemPromptBuilder — assemble the final system prompt from layered
// sources. Port of apps/server/agent/src/systemprompt/builder.ts.
//
// The prompt has two parts. StableSystem holds what never changes within a
// session and is sent as the system prompt: identity/SYSTEM.md,
// approval-mode instructions, what skills are for, tool inventory, the
// <critical> block, and any appended prompt. Setup holds the per-session
// context sent as a leading user message: skills list, rules, slash
// commands, repo context files (AGENTS.md, …), workspace tree, and
// workstation info (date/cwd/OS/git/model).
package systemprompt

import (
	"fmt"
	"strings"
)

// ApprovalMode mirrors ApprovalMode by value (not by import, to avoid a
// tools -> systemprompt -> permissions -> tools cycle: readSkill lives in
// tools and needs systemprompt, permissions needs tools.ToolTier).
type ApprovalMode string

const (
	AlwaysAsk   ApprovalMode = "always-ask"
	AcceptEdits ApprovalMode = "accept-edits"
	PlanMode    ApprovalMode = "plan-mode"
	FullAccess  ApprovalMode = "full-access"
)

const defaultIdentity = `You are a coding agent.
Use the available tools to explore the codebase, edit files, run shell commands, and accomplish user tasks correctly and concisely.`

// DefaultToolNames lists the built-in tool inventory shown when the caller
// doesn't supply an explicit set.
var DefaultToolNames = []string{
	"read_file", "readSkill", "list_dir", "glob", "grep",
	"write_file", "editFile", "batchWrite", "bash", "bashJob", "webSearch", "webFetch",
	"todo", "subagent", "ask", "askMany", "memory",
}

// DiscoveredContext is everything discovery found for one build.
type DiscoveredContext struct {
	ContextFiles     []ContextFile
	SystemPromptFile *SystemPromptFile
	Rules            []Rule
	Skills           []Skill
	Commands         []SlashCommand
	WorkspaceTree    WorkspaceTree
	Environment      EnvironmentInfo
}

// BuildOptions configures one discovery + prompt-assembly run. The
// workspace tree section is included by default; set SkipWorkspaceTree to
// omit it (matches the TS `includeWorkspaceTree: false` option).
type BuildOptions struct {
	Cwd               string
	Home              string
	StopAt            string
	Model             string
	ToolNames         []string
	ApprovalMode      ApprovalMode
	CustomPrompt      string
	AppendPrompt      string
	SkipWorkspaceTree bool
	Identity          string
}

// Result is the outcome of BuildSystemPrompt.
type Result struct {
	// SystemPrompt is StableSystem and Setup joined, for callers that send
	// everything as one system prompt.
	SystemPrompt string
	// StableSystem has no per-session values (date, cwd, branch, …), so it
	// is identical across sessions with the same mode and tools.
	StableSystem string
	// Setup is the per-session context, sent as a leading user message.
	Setup    string
	Sections []Section
	Context  DiscoveredContext
}

// Section is one named, non-empty block of the assembled prompt. Setup
// marks blocks that belong to Result.Setup rather than StableSystem.
type Section struct {
	Name    string
	Content string
	Setup   bool
}

func section(name, content string) *Section {
	trimmed := strings.TrimSpace(content)
	if trimmed == "" {
		return nil
	}
	return &Section{Name: name, Content: trimmed}
}

// DiscoverContext runs every discovery loader and assembles the aggregate
// context for one cwd/home/stopAt.
func DiscoverContext(opts BuildOptions) DiscoveredContext {
	roots := ResolveRoots(opts.Cwd, opts.Home, opts.StopAt)

	tree := WorkspaceTree{RootPath: roots.Cwd}
	if !opts.SkipWorkspaceTree {
		tree = BuildWorkspaceTree(roots.Cwd)
	}

	return DiscoveredContext{
		ContextFiles:     DiscoverContextFiles(roots),
		SystemPromptFile: DiscoverSystemPromptFile(roots),
		Rules:            DiscoverRules(roots),
		Skills:           DiscoverSkills(roots),
		Commands:         DiscoverCommands(roots),
		WorkspaceTree:    tree,
		Environment:      CollectEnvironmentInfo(roots.Cwd, opts.Model),
	}
}

func renderIdentity(opts BuildOptions, systemMd string) string {
	if strings.TrimSpace(opts.CustomPrompt) != "" {
		return strings.TrimSpace(opts.CustomPrompt)
	}
	if strings.TrimSpace(systemMd) != "" {
		return strings.TrimSpace(systemMd)
	}
	if strings.TrimSpace(opts.Identity) != "" {
		return strings.TrimSpace(opts.Identity)
	}
	return defaultIdentity
}

func renderSkills(skills []Skill) string {
	var lines []string
	for _, s := range skills {
		if s.Hide {
			continue
		}
		desc := strings.TrimSpace(s.Description)
		if desc == "" {
			desc = "(no description)"
		}
		lines = append(lines, fmt.Sprintf("- %s: %s", s.Name, desc))
	}
	if len(lines) == 0 {
		return ""
	}
	out := []string{"# Skills", "<skills>"}
	out = append(out, lines...)
	out = append(out, "</skills>")
	return strings.Join(out, "\n")
}

// renderSkillsGuide explains what skills are; the list itself is in setup.
func renderSkillsGuide(skills []Skill) string {
	for _, s := range skills {
		if !s.Hide {
			return "# Skills\nSkills are specialized knowledge. The available skills are listed in the setup message. When one matches the current task, call the `readSkill` tool with the skill name to load its full instructions before proceeding."
		}
	}
	return ""
}

func renderAlwaysApplyRules(rules []Rule) string {
	var blocks []string
	for _, r := range rules {
		if !r.AlwaysApply || strings.TrimSpace(r.Content) == "" {
			continue
		}
		blocks = append(blocks, fmt.Sprintf("<rule name=%q path=%q>\n%s\n</rule>", r.Name, r.Path, strings.TrimSpace(r.Content)))
	}
	if len(blocks) == 0 {
		return ""
	}
	out := []string{"# Always-apply rules", "<generic-rules>"}
	out = append(out, blocks...)
	out = append(out, "</generic-rules>")
	return strings.Join(out, "\n")
}

func renderDomainRules(rules []Rule) string {
	var lines []string
	for _, r := range rules {
		if r.AlwaysApply {
			continue
		}
		globs := ""
		if len(r.Globs) > 0 {
			globs = fmt.Sprintf(" (%s)", strings.Join(r.Globs, ", "))
		}
		desc := strings.TrimSpace(r.Description)
		if desc == "" {
			desc = "see rule file"
		}
		lines = append(lines, fmt.Sprintf("- %s%s: %s", r.Name, globs, desc))
	}
	if len(lines) == 0 {
		return ""
	}
	out := []string{"# Domain rules (request when relevant)", "<domain-rules>"}
	out = append(out, lines...)
	out = append(out, "</domain-rules>")
	return strings.Join(out, "\n")
}

func renderTools(names []string) string {
	if len(names) == 0 {
		return ""
	}
	lines := []string{"# Tool inventory"}
	for _, n := range names {
		lines = append(lines, fmt.Sprintf("- `%s`", n))
	}
	return strings.Join(lines, "\n")
}

func renderContextFiles(files []ContextFile) string {
	if len(files) == 0 {
		return ""
	}
	lines := []string{"# Repo rules", "You MUST follow the context files below for all tasks:", "<repo-rules>"}
	for _, f := range files {
		lines = append(lines, fmt.Sprintf("<file path=%q>\n%s\n</file>", f.Path, strings.TrimSpace(f.Content)))
	}
	lines = append(lines, "</repo-rules>",
		"Context files above are loaded automatically. Do not waste tools grepping for AGENTS.md / CLAUDE.md unless the user asks.")
	return strings.Join(lines, "\n")
}

func renderWorkspaceTree(tree WorkspaceTree, skip bool) string {
	if skip || strings.TrimSpace(tree.Rendered) == "" {
		return ""
	}
	return strings.Join([]string{
		"# Workspace tree",
		"<workspace-tree>",
		"Working directory layout (depth-limited):",
		tree.Rendered,
		"</workspace-tree>",
	}, "\n")
}

func renderWorkstation(env EnvironmentInfo) string {
	lines := []string{
		fmt.Sprintf("Date: %s", env.Date),
		fmt.Sprintf("CWD: %s", env.Cwd),
		fmt.Sprintf("OS: %s", env.OS),
		fmt.Sprintf("Arch: %s", env.Arch),
	}
	if env.GitBranch != "" {
		lines = append(lines, fmt.Sprintf("Git branch: %s", env.GitBranch))
	}
	if env.Model != "" {
		lines = append(lines, fmt.Sprintf("Model: %s", env.Model))
	}

	out := []string{"# Workstation", "<workstation>"}
	for _, l := range lines {
		out = append(out, "- "+l)
	}
	out = append(out, "</workstation>", fmt.Sprintf("Today is %s. Working directory is '%s'.", env.Date, env.Cwd))
	return strings.Join(out, "\n")
}

func renderCritical(mode ApprovalMode) string {
	if mode == PlanMode {
		return strings.Join([]string{
			"<critical>",
			"- Each response MUST advance the task.",
			"- Default to discussion and questions; do not write files or implement features.",
			"</critical>",
		}, "\n")
	}
	return strings.Join([]string{
		"<critical>",
		"- Each response MUST advance the task.",
		"- Default to informed action; do not ask for confirmation when tools or repo context can answer.",
		"</critical>",
	}, "\n")
}

func renderCommands(commands []SlashCommand) string {
	lines := []string{fmt.Sprintf("- /%s: %s. %s", InitCommandName, InitCommandDescription, InitCommandDetails)}
	for _, c := range commands {
		desc := strings.TrimSpace(c.Description)
		if desc == "" {
			desc = "user-defined command"
		}
		lines = append(lines, fmt.Sprintf("- /%s: %s", c.Name, desc))
	}
	out := []string{"# Available slash commands (may be invoked by the user)"}
	out = append(out, lines...)
	return strings.Join(out, "\n")
}

func renderApprovalModeInstruction(mode ApprovalMode) string {
	if mode == "" {
		return ""
	}
	heading := "# Approval Mode Active"
	tag := "<approval-mode-instructions>"

	switch mode {
	case AlwaysAsk:
		return strings.Join([]string{
			heading, tag,
			"You are in Normal mode. Use tools freely, including writes, edits, and command execution.",
			"Restricted actions trigger an approval dialog for the user automatically; never use the ask tool to request permission and never pause to ask permission yourself.",
			"Do not claim a restricted action was performed until it is approved.",
			"If approval is denied, stop and report the denial rather than retrying silently.",
			"</approval-mode-instructions>",
		}, "\n")
	case AcceptEdits:
		return strings.Join([]string{
			heading, tag,
			"You are in Accept Edits mode. Read, write, and edit files directly without pausing for each edit.",
			"Commands and higher-risk operations trigger an automatic approval dialog; never pause to ask permission yourself.",
			"Proceed efficiently through multi-file edits; do not stop to ask permission for routine file writes.",
			"</approval-mode-instructions>",
		}, "\n")
	case PlanMode:
		return strings.Join([]string{
			heading, tag,
			"Plan mode is ACTIVE. Read-only by default:",
			"- You may read files, search the codebase, and ask clarifying questions to shape the design.",
			"- You MUST NOT create, edit, or delete files; run build/install/migration/test commands; commit, push, or spawn implementation subagents.",
			"- Plans live in chat. Only write a plan file when the user explicitly asks.",
			"- If the user asks you to implement, tell them to switch agents. Do not silently start implementing.",
			"</approval-mode-instructions>",
		}, "\n")
	case FullAccess:
		return strings.Join([]string{
			heading, tag,
			"You are in Bypass Permissions mode. Execute all necessary tools directly without asking for approval.",
			"You remain responsible for avoiding unrelated or destructive work; freedom is not a license to be careless.",
			"Report every action you take accurately so the user can audit what happened.",
			"</approval-mode-instructions>",
		}, "\n")
	default:
		return ""
	}
}

// BuildSystemPrompt runs discovery then assembles the final prompt string.
func BuildSystemPrompt(opts BuildOptions) Result {
	if opts.ToolNames == nil {
		opts.ToolNames = DefaultToolNames
	}
	context := DiscoverContext(opts)
	identity := renderIdentity(opts, systemMdContent(context.SystemPromptFile))

	var stable, setup []Section
	add := func(dst *[]Section, s *Section) {
		if s != nil {
			*dst = append(*dst, *s)
		}
	}
	add(&stable, section("identity", identity))
	add(&stable, section("approval-mode", renderApprovalModeInstruction(opts.ApprovalMode)))
	add(&stable, section("skills-guide", renderSkillsGuide(context.Skills)))
	add(&stable, section("tools", renderTools(opts.ToolNames)))
	add(&stable, section("critical", renderCritical(opts.ApprovalMode)))
	add(&stable, section("append", opts.AppendPrompt))

	add(&setup, section("skills", renderSkills(context.Skills)))
	add(&setup, section("always-apply-rules", renderAlwaysApplyRules(context.Rules)))
	add(&setup, section("domain-rules", renderDomainRules(context.Rules)))
	add(&setup, section("commands", renderCommands(context.Commands)))
	add(&setup, section("repo-rules", renderContextFiles(context.ContextFiles)))
	add(&setup, section("workspace-tree", renderWorkspaceTree(context.WorkspaceTree, opts.SkipWorkspaceTree)))
	add(&setup, section("workstation", renderWorkstation(context.Environment)))
	for i := range setup {
		setup[i].Setup = true
	}

	stableText, setupText := joinSections(stable), joinSections(setup)
	full := stableText
	if setupText != "" {
		full = strings.TrimSpace(stableText + "\n\n" + setupText)
	}
	return Result{
		SystemPrompt: full,
		StableSystem: stableText,
		Setup:        setupText,
		Sections:     append(stable, setup...),
		Context:      context,
	}
}

func joinSections(sections []Section) string {
	joined := make([]string, len(sections))
	for i, p := range sections {
		joined[i] = p.Content
	}
	return strings.Join(joined, "\n\n")
}

func systemMdContent(f *SystemPromptFile) string {
	if f == nil {
		return ""
	}
	return f.Content
}
